src/operator/upsert.rs

```diff
@@ -0,0 +1,134 @@
+use crate::{
+    algebra::{HasOne, HasZero},
+    circuit::{
+        operator_traits::{BinaryOperator, Operator},
+        OwnershipPreference, Scope,
+    },
+    trace::{cursor::Cursor, Batch, Trace},
+};
+use std::{borrow::Cow, marker::PhantomData, mem::swap, ops::Neg};
+
+pub struct Upsert<K, VI, F, V, T, B> {
+    val_func: F,
+    phantom: PhantomData<(K, VI, V, T, B)>,
+}
+
+impl<K, VI, F, V, T, B> Upsert<K, VI, F, V, T, B> {
+    pub fn new(val_func: F) -> Self {
+        Self {
+            val_func,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<K, VI, F, V, T, B> Operator for Upsert<K, VI, F, V, T, B>
+where
+    B: 'static,
+    T: 'static,
+    V: 'static,
+    F: 'static,
+    VI: 'static,
+    K: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Upsert")
+    }
+
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+fn skip_zero_weights<'s, K, V, R, C>(cursor: &mut C) -> bool
+where
+    C: Cursor<'s, K, V, (), R>,
+    R: HasZero,
+{
+    while cursor.val_valid() && cursor.weight().is_zero() {
+        cursor.step_val();
+    }
+    cursor.val_valid()
+}
+
+impl<K, VI, F, V, T, B> BinaryOperator<T, Vec<(K, VI)>, B> for Upsert<K, VI, F, V, T, B>
+where
+    T: Trace<Key = K, Val = V, Batch = B, Time = (), R = B::R> + 'static,
+    B: Batch<Key = K, Val = V, Time = ()> + 'static,
+    F: Fn(VI) -> Option<V> + 'static,
+    K: Ord + Clone + 'static,
+    VI: Eq + 'static,
+    B::R: HasOne + Neg<Output = B::R> + 'static,
+    V: Eq + Clone + 'static,
+{
+    fn eval(&mut self, _trace: &T, _upserts: &Vec<(K, VI)>) -> B {
+        panic!("Upsert::eval(): cannot accept upserts by reference")
```

Comment:
It's not an error message, it's an assertion that shouldn't occur.

Reply:
I think that users can misuse the library, causing this panic to occur. So it will behave like an error message.
A panic should ideally happen only when some invariant that you are supposed to obey is not respected.
If something can be triggered from external calls, it is not an invariant.

Reply:
This is not a public API.  I wouldn't use `panic` if that was the case.